sys/kern/kern_exec.c

```diff
@@ -1655,54 +1697,69 @@ exec_copyout_strings(struct image_params *imgp)
 		execpath_len = 0;
 	p = imgp->proc;
 	szsigcode = 0;
+
+#if __has_feature(capabilities)
+	/* XXX: should be derived from a capability to the strings region */
+	arginfo = cheri_capability_build_user_data(
+	    CHERI_CAP_USER_DATA_PERMS, CHERI_CAP_USER_DATA_BASE,
+	    CHERI_CAP_USER_DATA_LENGTH,
+	    p->p_sysent->sv_psstrings);
+#else
 	arginfo = (struct ps_strings *)p->p_sysent->sv_psstrings;
+#endif
 	if (p->p_sysent->sv_sigcode_base == 0) {
 		if (p->p_sysent->sv_szsigcode != NULL)
 			szsigcode = *(p->p_sysent->sv_szsigcode);
 	}
-	destp =	(uintptr_t)arginfo;
+	destp =	(char * __capability)arginfo;
 
 	/*
 	 * install sigcode
 	 */
 	if (szsigcode != 0) {
 		destp -= szsigcode;
-		destp = rounddown2(destp, sizeof(void *));
-		copyout(p->p_sysent->sv_sigcode, (void *)destp, szsigcode);
+		destp = __builtin_align_down(destp,
+		    sizeof(void * __capability));
+		copyout_c(p->p_sysent->sv_sigcode, destp, szsigcode);
 	}
 
 	/*
 	 * Copy the image path for the rtld.
 	 */
 	if (execpath_len != 0) {
 		destp -= execpath_len;
-		destp = rounddown2(destp, sizeof(void *));
-		imgp->execpathp = destp;
-		copyout(imgp->execpath, (void *)destp, execpath_len);
+
+		/* XXX: CheriABI didn't do the align_down */
+		destp = __builtin_align_down(destp,
+		    sizeof(void * __capability));
```

Comment:
While PATH_MAX is small enough it doesn't matter, we should probably be padding as required for each thing that shouldn't overlap something else even if it can't happen in practice so we aren't bit by it at some later date.

No need to do that today though.